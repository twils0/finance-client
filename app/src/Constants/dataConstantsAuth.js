export const actionTypes = {
  SET_AUTHENTICATED: 'DATA_AUTH_SET_AUTHENTICATED',
  SET_REDIRECT_URL: 'DATA_AUTH_SET_REDIRECT_URL',
  SET_AUTH_STATUS: 'DATA_AUTH_SET_AUTH_STATUS',
  SET_CODE_TYPE: 'DATA_AUTH_SET_CODE_TYPE',
  RESET_AUTH_STATE: 'DATA_AUTH_RESET_AUTH_STATE',
};

export const statusNames = {
  LOGIN: 'authStatusLogin',
  SIGN_OUT_DEVICES: 'authStatusSignOutDevices',
  LOGIN_MFA: 'authStatusLoginMFA',
  SIGN_UP: 'authStatusSignUp',
  VERIFY_PHONE: 'authStatusVerifyPhone',
  VERIFY_PHONE_CODE: 'authStatusVerifyPhoneCode',
  VERIFY_EMAIL: 'authStatusVerifyEmail',
  VERIFY_EMAIL_LINK: 'authStatusVerifyEmailLink',
  FORGOT_PASSWORD: 'authStatusForgotPassword',
  RESET_PASSWORD: 'authStatusResetPassword',
  CHANGE_PASSWORD: 'authStatusChangePassword',
  DELETE_ACCOUNT: 'authStatusDeleteAccount',
  LOGOUT: 'authStatusLogout',
};

export const codeTypeNames = {
  VERIFY_PHONE: 'verifyPhone',
  VERIFY_EMAIL: 'verifyEmail',
  VERIFY_EMAIL_ADDITIONAL: 'verifyEmailAdditional',
};
