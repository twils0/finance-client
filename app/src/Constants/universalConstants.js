import stripeConfig from '../../../stripe.config.json';
import dbConfig from '../../../db.config.json';
import iexConfig from '../../../iex.config.json';

export const baseURLs = {
  STRIPE: stripeConfig.url,
  DB: dbConfig.url,
  IEX: iexConfig.urls.api,
  IEX_DEV: iexConfig.urls.dev,
  IEX_TERMS: iexConfig.urls.terms,
};

export const URLs = {
  CUSTOMERS: '/customers',
  USERS: '/users',
  EMAILS: '/emails',
  SECURITIES: '/securities',
  SECURITIES_SEARCH: '/securities/search',
  IEX_COMPANY: tickerCusip => `/stock/${tickerCusip}/company`,
  IEX_QUOTE: tickerCusip => `/stock/${tickerCusip}/quote`,
};

export const axiosConfig = {
  STRIPE: {
    baseURL: baseURLs.STRIPE,
    timeout: 30000,
  },
  DB: {
    baseURL: baseURLs.DB,
    timeout: 30000,
  },
  IEX: {
    baseURL: baseURLs.IEX,
    timeout: 30000,
  },
};

export const pathNames = {
  DEFAULT: '*',
  LOGIN: '/login',
  SIGN_UP: '/signup',
  CODE: '/code',
  CODE_VERIFY_EMAIL: '/code/:verificationId?',
  TERMS: '/terms',
  INTERNAL: '/internal',
  WATCHLIST: '/internal/watchlist',
  WATCHLIST_SECURITY_ID: '/internal/watchlist/:securityId?',
  ACCOUNT: '/internal/account',
};

export const requestStatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
};
