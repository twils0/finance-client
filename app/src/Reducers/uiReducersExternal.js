import { combineReducers } from 'redux';

import {
  formNames,
  inputNames,
  buttonNames,
  buttonTexts,
  actionTypes,
} from '../Constants/uiConstantsExternal';

export const initialState = {
  utilities: {
    forms: {
      springSettings: { stiffness: 170, damping: 26 },
      width: 300,
      widthBuffer: 60,
      mediaWidth: 700,
      heightBuffer: 50,
      inputs: {
        heightBuffer: 15,
      },
    },
    buttons: {
      width: 300,
      heightBuffer: 20,
      heightBottomBuffer: 30,
    },
    terms: {
      bufferPage: 30,
      bufferHeightParagraphTop: 20,
      bufferHeightParagraphBottom: 30,
      bufferHeight: 20,
    },
  },
  forms: {
    current: formNames.LOGIN,
    list: [...Object.values(formNames)],
    [formNames.RESET_PASSWORD]: {
      id: formNames.RESET_PASSWORD,
      height: 339,
      inputsList: [...Object.values(inputNames[formNames.RESET_PASSWORD])],
      inputs: {
        [inputNames[formNames.RESET_PASSWORD].CODE]: {
          parentId: formNames.RESET_PASSWORD,
          id: inputNames[formNames.RESET_PASSWORD].CODE,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.RESET_PASSWORD].PASSWORD]: {
          parentId: formNames.RESET_PASSWORD,
          id: inputNames[formNames.RESET_PASSWORD].PASSWORD,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.RESET_PASSWORD].PASSWORD2]: {
          parentId: formNames.RESET_PASSWORD,
          id: inputNames[formNames.RESET_PASSWORD].PASSWORD2,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.FORGOT_PASSWORD]: {
      id: formNames.FORGOT_PASSWORD,
      height: 220,
      inputsList: [...Object.values(inputNames[formNames.FORGOT_PASSWORD])],
      inputs: {
        [inputNames[formNames.FORGOT_PASSWORD].EMAIL]: {
          parentId: formNames.FORGOT_PASSWORD,
          id: inputNames[formNames.FORGOT_PASSWORD].EMAIL,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.LOGIN]: {
      id: formNames.LOGIN,
      height: 222,
      inputsList: [...Object.values(inputNames[formNames.LOGIN])],
      inputs: {
        [inputNames[formNames.LOGIN].EMAIL]: {
          parentId: formNames.LOGIN,
          id: inputNames[formNames.LOGIN].EMAIL,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.LOGIN].PASSWORD]: {
          parentId: formNames.LOGIN,
          id: inputNames[formNames.LOGIN].PASSWORD,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.CARD]: {
      id: formNames.CARD,
      height: 360,
      inputsList: [...Object.values(inputNames[formNames.CARD])],
      inputs: {
        [inputNames[formNames.CARD].NAME_ON_CARD]: {
          parentId: formNames.CARD,
          id: inputNames[formNames.CARD].NAME_ON_CARD,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.CARD].PROMO_CODE]: {
          parentId: formNames.CARD,
          id: inputNames[formNames.CARD].PROMO_CODE,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.CARD].STRIPE]: {
          parentId: formNames.CARD,
          id: inputNames[formNames.CARD].STRIPE,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.INFO]: {
      id: formNames.INFO,
      height: 280,
      inputsList: [...Object.values(inputNames[formNames.INFO])],
      inputs: {
        [inputNames[formNames.INFO].NAME]: {
          parentId: formNames.INFO,
          id: inputNames[formNames.INFO].NAME,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.INFO].EMAIL]: {
          parentId: formNames.INFO,
          id: inputNames[formNames.INFO].EMAIL,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.INFO].PHONE]: {
          parentId: formNames.INFO,
          id: inputNames[formNames.INFO].PHONE,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.SIGN_UP]: {
      id: formNames.SIGN_UP,
      height: 247,
      inputsList: [...Object.values(inputNames[formNames.SIGN_UP])],
      inputs: {
        [inputNames[formNames.SIGN_UP].PASSWORD]: {
          parentId: formNames.SIGN_UP,
          id: inputNames[formNames.SIGN_UP].PASSWORD,
          value: '',
          errorMessage: '',
        },
        [inputNames[formNames.SIGN_UP].PASSWORD2]: {
          parentId: formNames.SIGN_UP,
          id: inputNames[formNames.SIGN_UP].PASSWORD2,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.CODE_MFA_PHONE]: {
      id: formNames.CODE_MFA_PHONE,
      height: 183,
      inputsList: [...Object.values(inputNames[formNames.CODE_MFA_PHONE])],
      inputs: {
        [inputNames[formNames.CODE_MFA_PHONE].CODE]: {
          parentId: formNames.CODE_MFA_PHONE,
          id: inputNames[formNames.CODE_MFA_PHONE].CODE,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.CODE_VERIFY_PHONE]: {
      id: formNames.CODE_VERIFY_PHONE,
      height: 183,
      inputsList: [...Object.values(inputNames[formNames.CODE_VERIFY_PHONE])],
      inputs: {
        [inputNames[formNames.CODE_VERIFY_PHONE].CODE]: {
          parentId: formNames.CODE_VERIFY_PHONE,
          id: inputNames[formNames.CODE_VERIFY_PHONE].CODE,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.CODE_VERIFY_EMAIL]: {
      id: formNames.CODE_VERIFY_EMAIL,
      height: 183,
      inputsList: [...Object.values(inputNames[formNames.CODE_VERIFY_EMAIL])],
      inputs: {
        [inputNames[formNames.CODE_VERIFY_EMAIL].CODE]: {
          parentId: formNames.CODE_VERIFY_EMAIL,
          id: inputNames[formNames.CODE_VERIFY_EMAIL].CODE,
          value: '',
          errorMessage: '',
        },
      },
    },
    [formNames.DEVICE]: {
      id: formNames.DEVICE,
      height: 146,
      inputsList: [],
      inputs: {},
    },
  },
  buttons: {
    list: [...Object.values(buttonNames)],
    [buttonNames.FIRST]: { id: buttonNames.FIRST, text: buttonTexts.LOG_IN },
    [buttonNames.SECOND]: { id: buttonNames.SECOND, text: buttonTexts.SIGN_UP },
    [buttonNames.FORGOT_PASSWORD]: {
      id: buttonNames.FORGOT_PASSWORD,
      text: buttonTexts.FORGOT_PASSWORD,
    },
    [buttonNames.RESEND_PHONE]: { id: buttonNames.RESEND_PHONE, text: buttonTexts.RESEND_PHONE },
    [buttonNames.RESEND_EMAIL]: { id: buttonNames.RESEND_EMAIL, text: buttonTexts.RESEND_EMAIL },
  },
};

export const utilities = (state = initialState.utilities) => state;

export const forms = (state = initialState.forms, action) => {
  switch (action.type) {
    case actionTypes.SET_CURRENT_FORM:
      return {
        ...state,
        ...action.payload,
      };
    case actionTypes.SET_FORM_HEIGHT:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.SET_INPUT_VALUE_ERROR:
      return {
        ...state,
        [action.payload.parentId]: {
          ...state[action.payload.parentId],
          inputs: {
            ...state[action.payload.parentId].inputs,
            [action.payload.id]: {
              ...state[action.payload.parentId].inputs[action.payload.id],
              ...action.payload,
            },
          },
        },
      };
    default:
      return state;
  }
};

export const buttons = (state = initialState.buttons, action) => {
  switch (action.type) {
    case actionTypes.SET_BUTTON_TEXT:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    default:
      return state;
  }
};

const uiReducersExternal = combineReducers({
  utilities,
  forms,
  buttons,
});

export default uiReducersExternal;
