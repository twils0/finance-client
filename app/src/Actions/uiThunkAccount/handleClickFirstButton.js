import { formNames } from '../../Constants/uiConstantsAccount';

import firstButtonCode from './firstButtonCode';
import firstButtonProfile from './firstButtonProfile';
import firstButtonBilling from './firstButtonBilling';
import firstButtonChangePassword from './firstButtonChangePassword';
import firstButtonDeleteAccount from './firstButtonDeleteAccount';

const handleClickFirstButton = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'stripeElement')) {
    throw new Error(`Please enter a value for the 'stripeElement' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'resetStripeElement')) {
    throw new Error(`Please enter a value for the 'resetStripeElement' key - ${JSON.stringify(payload)}`);
  }

  return (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.internal.account;

    switch (forms.current) {
      case formNames.CODE:
        dispatch(firstButtonCode());
        break;
      case formNames.PROFILE:
        dispatch(firstButtonProfile());
        break;
      case formNames.BILLING: {
        const { stripeElement, resetStripeElement } = payload;

        dispatch(firstButtonBilling({ stripeElement, resetStripeElement }));
        break;
      }
      case formNames.CHANGE_PASSWORD:
        dispatch(firstButtonChangePassword());
        break;
      case formNames.DELETE_ACCOUNT: {
        dispatch(firstButtonDeleteAccount());
        break;
      }
      default:
        break;
    }

    return null;
  };
};

export default handleClickFirstButton;
