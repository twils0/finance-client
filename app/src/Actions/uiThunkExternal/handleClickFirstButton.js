import { formNames } from '../../Constants/uiConstantsExternal';
import { setCurrentForm } from '../uiActionsExternal';

import firstButtonResetPassword from './firstButtonResetPassword';
import firstButtonForgotPassword from './firstButtonForgotPassword';
import firstButtonLogin from './firstButtonLogin';
import firstButtonSignUp from './firstButtonSignUp';
import firstButtonCodeMFAPhone from './firstButtonCodeMFAPhone';
import firstButtonCodeVerifyPhone from './firstButtonCodeVerifyPhone';
import firstButtonCodeVerifyEmail from './firstButtonCodeVerifyEmail';
import firstButtonDevice from './firstButtonDevice';

const handleClickFirstButton = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'history')) {
    throw new Error(`Please enter a value for the 'history' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'match')) {
    throw new Error(`Please enter a value for the 'match' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'stripeElement')) {
    throw new Error(
      `Please enter a value for the 'stripeElement' key - ${JSON.stringify(payload)}`,
    );
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'resetStripeElement')) {
    throw new Error(
      `Please enter a value for the 'resetStripeElement' key - ${JSON.stringify(payload)}`,
    );
  }

  return (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.external;

    switch (forms.current) {
      case formNames.RESET_PASSWORD:
        dispatch(firstButtonResetPassword());
        break;
      case formNames.FORGOT_PASSWORD: {
        const { history } = payload;

        dispatch(firstButtonForgotPassword({ history }));
        break;
      }
      case formNames.LOGIN: {
        const { history } = payload;

        dispatch(firstButtonLogin({ history }));
        break;
      }
      case formNames.CARD:
        dispatch(setCurrentForm({ current: formNames.INFO }));
        break;
      case formNames.INFO:
        dispatch(setCurrentForm({ current: formNames.SIGN_UP }));
        break;
      case formNames.SIGN_UP: {
        const { history, stripeElement, resetStripeElement } = payload;

        dispatch(firstButtonSignUp({ history, stripeElement, resetStripeElement }));
        break;
      }
      case formNames.CODE_MFA_PHONE: {
        dispatch(firstButtonCodeMFAPhone());
        break;
      }
      case formNames.CODE_VERIFY_PHONE: {
        const { history } = payload;

        dispatch(firstButtonCodeVerifyPhone({ history }));
        break;
      }
      case formNames.CODE_VERIFY_EMAIL: {
        const { history, match } = payload;

        dispatch(firstButtonCodeVerifyEmail({ history, match }));
        break;
      }
      case formNames.DEVICE: {
        const { history } = payload;

        dispatch(firstButtonDevice({ history }));
        break;
      }
      default:
        break;
    }

    return null;
  };
};

export default handleClickFirstButton;
