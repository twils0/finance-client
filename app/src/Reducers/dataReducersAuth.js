import { combineReducers } from 'redux';

import { requestStatusTypes, pathNames } from '../Constants/universalConstants';
import { actionTypes, statusNames, codeTypeNames } from '../Constants/dataConstantsAuth';

export const initialState = {
  authenticated: false,
  redirectURL: pathNames.WATCHLIST,
  status: {
    [statusNames.LOGIN]: {
      id: statusNames.LOGIN,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_OUT_DEVICES]: {
      id: statusNames.SIGN_OUT_DEVICES,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGIN_MFA]: {
      id: statusNames.LOGIN_MFA,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_UP]: {
      id: statusNames.SIGN_UP,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE]: {
      id: statusNames.VERIFY_PHONE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE_CODE]: {
      id: statusNames.VERIFY_PHONE_CODE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL]: {
      id: statusNames.VERIFY_EMAIL,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL_LINK]: {
      id: statusNames.VERIFY_EMAIL_LINK,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.FORGOT_PASSWORD]: {
      id: statusNames.FORGOT_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.RESET_PASSWORD]: {
      id: statusNames.RESET_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.CHANGE_PASSWORD]: {
      id: statusNames.CHANGE_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGOUT]: {
      id: statusNames.LOGOUT,
      status: requestStatusTypes.SUCCESS,
    },
    [statusNames.DELETE_ACCOUNT]: {
      id: statusNames.DELETE_ACCOUNT,
      status: requestStatusTypes.SUCCESS,
    },
  },
  codeTypes: {
    [codeTypeNames.VERIFY_PHONE]: {
      id: codeTypeNames.VERIFY_PHONE,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL]: {
      id: codeTypeNames.VERIFY_EMAIL,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL_ADDITIONAL]: {
      id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL,
      needed: true,
    },
  },
};

export const authenticated = (state = initialState.authenticated, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTHENTICATED:
      return action.payload.authenticated;
    case actionTypes.RESET_AUTH_STATE:
      return initialState.authenticated;
    default:
      return state;
  }
};

export const redirectURL = (state = initialState.redirectURL, action) => {
  switch (action.type) {
    case actionTypes.SET_REDIRECT_URL:
      return action.payload.redirectURL;
    default:
      return state;
  }
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTH_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const codeTypes = (state = initialState.codeTypes, action) => {
  switch (action.type) {
    case actionTypes.SET_CODE_TYPE:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.codeTypes;
    default:
      return state;
  }
};

const dataReducersAuth = combineReducers({
  authenticated,
  redirectURL,
  status,
  codeTypes,
});

export default dataReducersAuth;
