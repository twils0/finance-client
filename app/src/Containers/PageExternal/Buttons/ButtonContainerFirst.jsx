import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import { requestStatusTypes } from '../../../Constants/universalConstants';
import { statusNames } from '../../../Constants/dataConstantsAuth';
import { buttonNames } from '../../../Constants/uiConstantsExternal';

import handleClickFirstButton from '../../../Actions/uiThunkExternal/handleClickFirstButton';

import ButtonFirst from '../../../Components/PageExternal/Buttons/ButtonFirst';

class ButtonContainerFirst extends React.Component {
  handleClick = (clickEvent) => {
    clickEvent.preventDefault();
    const {
      handleClickButton, history, match, stripeElement, resetStripeElement,
    } = this.props;

    handleClickButton({
      history,
      match,
      stripeElement,
      resetStripeElement,
    });
  };

  render() {
    const {
      width, heightBuffer, statusAuth, text,
    } = this.props;
    let disabled = false;

    if (
      statusAuth[statusNames.LOGIN].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.SIGN_OUT_DEVICES].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.LOGIN_MFA].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.VERIFY_PHONE].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.VERIFY_PHONE_CODE].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.VERIFY_EMAIL].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.FORGOT_PASSWORD].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.CHANGE_PASSWORD].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.SIGN_UP].status === requestStatusTypes.LOADING
      || statusAuth[statusNames.SIGN_UP].status === requestStatusTypes.SUCCESS
    ) {
      disabled = true;
    }

    return (
      <ButtonFirst
        width={width}
        heightBuffer={heightBuffer}
        disabled={disabled}
        text={text}
        handleClick={this.handleClick}
      />
    );
  }
}

ButtonContainerFirst.propTypes = {
  history: PropTypes.object.isRequired,
  match: PropTypes.object.isRequired,
  stripeElement: PropTypes.object,
  width: PropTypes.number.isRequired,
  heightBuffer: PropTypes.number.isRequired,
  statusAuth: PropTypes.object.isRequired,
  text: PropTypes.string.isRequired,
  resetStripeElement: PropTypes.func.isRequired,
  handleClickButton: PropTypes.func.isRequired,
};

ButtonContainerFirst.defaultProps = {
  stripeElement: {},
};

const mapStateToProps = (state, ownProps) => ({
  history: ownProps.history,
  match: ownProps.match,
  stripeElement: ownProps.stripeElement,
  width: state.ui.external.utilities.buttons.width,
  heightBuffer: state.ui.external.utilities.buttons.heightBuffer,
  statusAuth: state.data.auth.status,
  text: state.ui.external.buttons[buttonNames.FIRST].text,
  resetStripeElement: ownProps.resetStripeElement,
});

const mapDispatchToProps = dispatch => ({
  handleClickButton: payload => dispatch(handleClickFirstButton(payload)),
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(ButtonContainerFirst),
);
