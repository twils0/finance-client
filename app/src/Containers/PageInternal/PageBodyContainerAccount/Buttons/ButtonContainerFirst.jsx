import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { requestStatusTypes } from '../../../../Constants/universalConstants';
import { buttonNames } from '../../../../Constants/uiConstantsAccount';
import { statusNames } from '../../../../Constants/dataConstantsAccount';
import handleClickFirstButton from '../../../../Actions/uiThunkAccount/handleClickFirstButton';

import ButtonFirst from '../../../../Components/PageInternal/PageBodyAccount/Buttons/ButtonFirst';

class ButtonContainerFirst extends React.Component {
  handleClick = (clickEvent) => {
    clickEvent.preventDefault();
    const { handleClickButton, stripeElement, resetStripeElement } = this.props;

    handleClickButton({ stripeElement, resetStripeElement });
  };

  render() {
    const {
      statusAccount, width, heightBuffer, firstButtonText,
    } = this.props;
    let disabled = false;

    if (
      statusAccount[statusNames.UPDATE_STRIPE_FIELDS_REQUEST].status ===
        requestStatusTypes.LOADING ||
      statusAccount[statusNames.UPDATE_DB_FIELDS].status === requestStatusTypes.LOADING ||
      statusAccount[statusNames.UPDATE_AWS_FIELDS].status === requestStatusTypes.LOADING
    ) {
      disabled = true;
    }

    return (
      <ButtonFirst
        width={width}
        heightBuffer={heightBuffer}
        disabled={disabled}
        text={firstButtonText}
        handleClick={this.handleClick}
      />
    );
  }
}

ButtonContainerFirst.propTypes = {
  width: PropTypes.number.isRequired,
  heightBuffer: PropTypes.number.isRequired,
  statusAccount: PropTypes.object.isRequired,
  firstButtonText: PropTypes.string.isRequired,
  stripeElement: PropTypes.object,
  resetStripeElement: PropTypes.func.isRequired,
  handleClickButton: PropTypes.func.isRequired,
};

ButtonContainerFirst.defaultProps = {
  stripeElement: {},
};

const mapStateToProps = (state, ownProps) => ({
  width: state.ui.internal.account.utilities.buttons.width,
  heightBuffer: state.ui.internal.account.utilities.buttons.heightBuffer,
  statusAccount: state.data.account.status,
  firstButtonText: state.ui.internal.account.buttons[buttonNames.FIRST].text,
  stripeElement: ownProps.stripeElement,
  resetStripeElement: ownProps.resetStripeElement,
});

const mapDispatchToProps = dispatch => ({
  handleClickButton: payload => dispatch(handleClickFirstButton(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ButtonContainerFirst);
