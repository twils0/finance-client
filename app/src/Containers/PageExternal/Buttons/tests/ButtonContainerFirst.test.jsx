import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';

import { initialState as initialStateAuth } from '../../../../Reducers/dataReducersAuth';
import { initialState as initialStateUIExternal } from '../../../../Reducers/uiReducersExternal';
import { requestStatusTypes } from '../../../../Constants/universalConstants';
import { statusNames } from '../../../../Constants/dataConstantsAuth';

import ButtonContainerFirst from '../ButtonContainerFirst';

import handleClickFirstButton from '../../../../Actions/uiThunkExternal/handleClickFirstButton';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

jest.mock('../../../../Actions/uiThunkExternal/handleClickFirstButton', () => jest.fn());
handleClickFirstButton.mockReturnValue(() => null);

const history = { test: 'testHistory' };
const match = { test: 'testMatch ' };

const clickEvent = {
  preventDefault: jest.fn(),
};
const stripeElement = { test: 'testStripeElement' };
const resetStripeElement = jest.fn();

const shallowComponent = store => shallow(
  <ButtonContainerFirst stripeElement={stripeElement} resetStripeElement={resetStripeElement} />,
  {
    context: {
      store,
      router: {
        history,
        route: {
          location: {},
          match,
        },
      },
    },
  },
)
  .dive()
  .dive()
  .dive();

describe('Containers', () => {
  describe('PageExternal', () => {
    describe('Buttons', () => {
      describe('ButtonContainerFirst', () => {
        afterEach(() => {
          handleClickFirstButton.mockReset();
        });

        it('shallow renders correctly and button works, with disabled false', async () => {
          const store = mockStore({
            data: {
              auth: initialStateAuth,
            },
            ui: {
              external: initialStateUIExternal,
            },
          });

          const wrapper = shallowComponent(store);

          const instance = wrapper.instance();
          const props = wrapper.props();

          instance.handleClick(clickEvent);

          expect(wrapper).toMatchSnapshot();
          expect(props.disabled).toEqual(false);
          expect(handleClickFirstButton).toBeCalledWith({
            history,
            match,
            stripeElement,
            resetStripeElement,
          });
        });

        it('shallow renders correctly, with disabled true', async () => {
          const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
          stateBeforeAuth.status[statusNames.SIGN_UP].status = requestStatusTypes.LOADING;

          const store = mockStore({
            data: {
              auth: stateBeforeAuth,
            },
            ui: {
              external: initialStateUIExternal,
            },
          });

          const wrapper = shallowComponent(store);

          const props = wrapper.props();

          expect(wrapper).toMatchSnapshot();
          expect(props.disabled).toEqual(true);
        });
      });
    });
  });
});
