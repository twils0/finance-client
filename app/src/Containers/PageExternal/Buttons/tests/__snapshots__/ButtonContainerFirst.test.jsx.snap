// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers PageExternal Buttons ButtonContainerFirst shallow renders correctly and button works, with disabled false 1`] = `
<ButtonFirst
  disabled={false}
  handleClick={[Function]}
  heightBuffer={20}
  text="Log In"
  width={300}
/>
`;

exports[`Containers PageExternal Buttons ButtonContainerFirst shallow renders correctly, with disabled true 1`] = `
<ButtonFirst
  disabled={true}
  handleClick={[Function]}
  heightBuffer={20}
  text="Log In"
  width={300}
/>
`;
